modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/util/ProjectedTupleTest.java

```diff
@@ -0,0 +1,175 @@
+
+        LOG.info("Seed is " + seed);
+
+        var builder = new BinaryTupleBuilder(ALL_TYPES_SCHEMA.elementCount());
+
+        for (int i = 0; i < ALL_TYPES_SCHEMA.elementCount(); i++) {
+            Element e = ALL_TYPES_SCHEMA.element(i);
+
+            BinaryRowConverter.appendValue(builder, e, SchemaTestUtils.generateRandomValue(RND, fromElement(e)));
+        }
+
+        TUPLE = new BinaryTuple(ALL_TYPES_SCHEMA.elementCount(), builder.build());
+    }
+
+    @Test
+    void allTypesAreCovered() {
+        List<NativeTypeSpec> coveredTypes = IntStream.range(0, ALL_TYPES_SCHEMA.elementCount())
+                .mapToObj(ALL_TYPES_SCHEMA::element)
+                .map(Element::typeSpec)
+                .collect(Collectors.toList());
+
+        EnumSet<NativeTypeSpec> allTypes = EnumSet.allOf(NativeTypeSpec.class);
+
+        coveredTypes.forEach(allTypes::remove);
+
+        assertThat(allTypes, empty());
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 2, 3})
+    void projectionReturnsProperElementCount(int projectionSize) {
+        InternalTuple projection1 = new ProjectedTuple(
+                ALL_TYPES_SCHEMA, TUPLE, new int[projectionSize]
+        );
+
+        assertThat(projection1.elementCount(), equalTo(projectionSize));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testProjection(boolean useOptimizeProjection) {
+        int f1 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+        int f2 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+        int f3 = RND.nextInt(ALL_TYPES_SCHEMA.elementCount());
+
+        int[] projection = {f1, f2, f3};
+
+        InternalTuple projectedTuple = useOptimizeProjection
+                ? new ProjectedTuple(TUPLE, projection)
+                : new ProjectedTuple(ALL_TYPES_SCHEMA, TUPLE, projection);
+
+        Element e1 = ALL_TYPES_SCHEMA.element(f1);
+        Element e2 = ALL_TYPES_SCHEMA.element(f2);
+        Element e3 = ALL_TYPES_SCHEMA.element(f3);
+
+        BinaryTupleSchema projectedSchema = BinaryTupleSchema.create(new Element[] {
+                e1, e2, e3
+        });
+
+        assertThat(projectedSchema.value(projectedTuple, 0), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f1)));
+        assertThat(projectedSchema.value(projectedTuple, 1), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f2)));
+        assertThat(projectedSchema.value(projectedTuple, 2), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f3)));
+
+        InternalTuple restored = new BinaryTuple(projection.length, projectedTuple.byteBuffer());
+
+        assertThat(projectedSchema.value(restored, 0), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f1)));
+        assertThat(projectedSchema.value(restored, 1), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f2)));
+        assertThat(projectedSchema.value(restored, 2), equalTo(ALL_TYPES_SCHEMA.value(TUPLE, f3)));
+    }
+
+    private static NativeType fromElement(Element element) {
+        switch (element.typeSpec()) {
```

Comment:
I see two other test classes that implement such a method. `AbstractSerializerTest#specToType` and `AbstractSchemaConverterTest#specToType`.

Seems (maybe someday in the future :roll_eyes:) we'll need to extract it into a common utility class (`SchemaTestUtils`) but provide ability to pass decimal precision/scale arguments :thinking:

Reply:
fair enough, but this is out of scope of this particular patch